lib/charms/haproxy/v0/haproxy_route_tcp.py

```diff
@@ -0,0 +1,1738 @@
+        # the relation IDs that request overlapping ports to ignore them during
+        # rendering of the haproxy configuration.
+        port_to_relation_ids: dict[int, list[int]] = {}
+        for requirer_data in self.requirers_data:
+            if not port_to_relation_ids.get(requirer_data.application_data.port):
+                port_to_relation_ids[requirer_data.application_data.port] = [
+                    requirer_data.relation_id
+                ]
+                continue
+            port_to_relation_ids[requirer_data.application_data.port].append(
+                requirer_data.relation_id
+            )
+
+        for relation_ids in port_to_relation_ids.values():
+            if len(relation_ids) > 1:
+                self.relation_ids_with_invalid_data.extend(relation_ids)
+        return self
+
+
+class HaproxyRouteTcpDataAvailableEvent(EventBase):
+    """HaproxyRouteDataAvailableEvent custom event.
+
+    This event indicates that the requirers data are available.
+    """
+
+
+class HaproxyRouteTcpDataRemovedEvent(EventBase):
+    """HaproxyRouteDataRemovedEvent custom event.
+
+    This event indicates that one of the endpoints was removed.
+    """
+
+
+class HaproxyRouteTcpProviderEvents(CharmEvents):
+    """List of events that the TLS Certificates requirer charm can leverage.
+
+    Attributes:
+        data_available: This event indicates that
+            the haproxy-route endpoints are available.
+        data_removed: This event indicates that one of the endpoints was removed.
+    """
+
+    data_available = EventSource(HaproxyRouteTcpDataAvailableEvent)
+    data_removed = EventSource(HaproxyRouteTcpDataRemovedEvent)
+
+
+class HaproxyRouteTcpProvider(Object):
+    """Haproxy-route interface provider implementation.
+
+    Attributes:
+        on: Custom events of the provider.
+        relations: Related appliations.
+    """
+
+    on = HaproxyRouteTcpProviderEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = HAPROXY_ROUTE_TCP_RELATION_NAME,
+        raise_on_validation_error: bool = False,
+    ) -> None:
+        """Initialize the HaproxyRouteProvider.
+
+        Args:
+            charm: The charm that is instantiating the library.
+            relation_name: The name of the relation.
+            raise_on_validation_error: Whether the library should raise
+                HaproxyRouteTcpInvalidRelationDataError when requirer data validation fails.
+                If this is set to True the provider charm needs to also catch and handle the
+                thrown exception.
+        """
+        super().__init__(charm, relation_name)
+
+        self._relation_name = relation_name
+        self.charm = charm
+        self.raise_on_validation_error = raise_on_validation_error
+        on = self.charm.on
+        self.framework.observe(on[self._relation_name].relation_created, self._configure)
```

Comment:
Won't this lead to duplicated ready events when the relation is created? Iirc both a relation created and a relation changes will be triggered